rasterio/rio/warp.py

```diff
@@ -1,66 +1,94 @@
 import logging
 from math import ceil
+import warnings
+
 import click
-from cligj import format_opt
+from cligj import files_inout_arg, format_opt
 
+from .helpers import resolve_inout
 from . import options
 import rasterio
 from rasterio import crs
 from rasterio.transform import Affine
-from rasterio.warp import (reproject, RESAMPLING, calculate_default_transform,
+from rasterio.warp import (reproject, Resampling, calculate_default_transform,
    transform_bounds)
 
 
-logger = logging.getLogger('rio')
+# Improper usage of rio-warp can lead to accidental creation of
+# extremely large datasets. We'll put a hard limit on the size of
+# datasets and raise a usage error if the limits are exceeded.
+MAX_OUTPUT_WIDTH = 100000
+MAX_OUTPUT_HEIGHT = 100000
```

Comment:
That'd be what - an 80GB in-memory raster with a single uint8 band? I think it's safe to say most rasterio users will never hit that limit :-)

But how do we future proof this? Could we estimate the max size based on the available system memory?

Bigger picture there are other cases like a [boundless read with the wrong crs](https://github.com/perrygeo/python-rasterstats/issues/107#issuecomment-166985520) that can cause ridiculous sized  numpy arrays or in-memory rasters. Do we want to do more at the _base level to gaurd against these cases?